Expose height and width factor in AnimatedAlign

Nash0x7E2 opened the pull request:
## Description
`Align` allows developers to customize the width and height factor, however `AnimatedAlign` does not expose these properties. This PR exposes `widthFactor` and `heightFactor` for AnimatedAlign and adds the necessary Tween visitors and documentation for both properties.

## Tests
No test added. I looked at the existing implementation of `Align` and saw that there was no test covering the height and width factors so none was added to the Animated widget. 

## Checklist

- [ x ] I read the [Contributor Guide] and followed the process outlined there for submitting PRs.
- [ x ] I signed the [CLA].
- [ x ] I read and followed the [Flutter Style Guide], including [Features we expect every widget to implement].
- [ x ] I read the [Tree Hygiene] wiki page, which explains my responsibilities.
- [ x ] I updated/added relevant documentation (doc comments with `///`).
- [ x ] All existing and new tests are passing.
- [ x ] The analyzer (`flutter analyze --flutter-repo`) does not report any problems on my PR.
- [ x ] I am willing to follow-up on review comments in a timely manner.

## Breaking Change

- [ x ] No, no existing tests failed, so this is *not* a breaking change.

Nash0x7E2 committed:
Expose height and width factor

- Expose the height and width factor properties in AnimatedAlign

Nash0x7E2 committed:
Add type information to tween

Piinks commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -941,6 +962,8 @@ class _AnimatedAlignState extends AnimatedWidgetBaseState<AnimatedAlign> {
   void debugFillProperties(DiagnosticPropertiesBuilder description) {
     super.debugFillProperties(description);
     description.add(DiagnosticsProperty<AlignmentGeometryTween>('alignment', _alignment, defaultValue: null));
+    description.add(DiagnosticsProperty<Tween>('width factor', _widthFactorTween, defaultValue: null));
```
```suggestion
    description.add(DiagnosticsProperty<Tween>('widthFactor', _widthFactorTween, defaultValue: null));
```

Piinks commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -941,6 +962,8 @@ class _AnimatedAlignState extends AnimatedWidgetBaseState<AnimatedAlign> {
     super.debugFillProperties(description);
     description.add(DiagnosticsProperty<AlignmentGeometryTween>('alignment', _alignment, defaultValue: null));
+    description.add(DiagnosticsProperty<Tween>('width factor', _widthFactorTween, defaultValue: null));
+    description.add(DiagnosticsProperty<Tween>('height factor', _heightFactorTween, defaultValue: null));
```
```suggestion
    description.add(DiagnosticsProperty<Tween>('heightFactor', _heightFactorTween, defaultValue: null));
```

Nash0x7E2 replied:
Oops 😆

Piinks commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -911,6 +915,16 @@ class AnimatedAlign extends ImplicitlyAnimatedWidget {
 
+  /// If non-null, sets its height to the child's height multiplied by this factor.
+  ///
+  /// Can be both greater and less than 1.0 but must be positive.
```
I may be clearer here and below to say Must be greater than or equal to 0.0, defaults to 1.0.

Nash0x7E2 replied:
Good point.

Piinks reviewed (commented):
Hey @Nash0x7E2! Thank you for the contribution! This will need tests. The fact that the Align class does not have tests covering these values is something we should fix. 😅

Nash0x7E2 commented:
Hey @Piinks! Thanks for the review! I'll add the tests later today 🙂

Nash0x7E2 committed:
add review suggestions

Nash0x7E2 committed:
add test for animated align

Nash0x7E2 committed:
add height/width factor test to align

Nash0x7E2 commented:
@Piinks I've updated the PR with your suggestions and the appropriate test for `AnimatedAlign`. 

I also added the missing test to the regular `Align` :)

Piinks commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -911,6 +915,16 @@ class AnimatedAlign extends ImplicitlyAnimatedWidget {
 
+  /// If non-null, sets its height to the child's height multiplied by this factor.
+  ///
+  /// Must be within the range 0.0 - 1.0. Defaults to 1.0.
```
I thought these could be greater than 1.0?

Nash0x7E2 replied:
They can 😬  Just pushed the correction :)

Piinks reviewed (commented):
Awesome! Thanks for going above and beyond and adding tests to the Align class! Just a doc nit below. Great stuff! :)

Nash0x7E2 committed:
minor doc update

Piinks reviewed (approved):
Great! LGTM! Thank you! 🎉

Nash0x7E2 commented:
Wooohoo 🎉

Piinks commented on packages/flutter/lib/src/widgets/implicit_animations.dart:
```diff
@@ -882,10 +882,14 @@ class AnimatedAlign extends ImplicitlyAnimatedWidget {
     @required Duration duration,
     VoidCallback onEnd,
   }) : assert(alignment != null),
+        assert(widthFactor == null || widthFactor >= 0.0),
+        assert(heightFactor == null || heightFactor >= 0.0),
```
@hansmuller just astutely noticed the Align-ment (😆 ) here is off. Can you fix this nit before we merge?

```suggestion
       assert(widthFactor == null || widthFactor >= 0.0),
       assert(heightFactor == null || heightFactor >= 0.0),
```

Nash0x7E2 replied:
Hahaha 😆  done

Nash0x7E2 committed:
Fix indentation :)

Co-authored-by: Kate Lovett <katelovett@google.com>
